File: src/gui/laanimalmanager.cpp
/***************************************************************************
 *   Copyright (C) 2007 by: Jason Jorgenson   arkygeek@gmail.com           *
 *             (c) 2007 by: Tim Sutton tim@linfiniti.com
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "laanimalmanager.h"
#include "lautils.h"
#include <QSettings>
#include <QTreeWidget>
#include <QTreeWidgetItem>
#include <QTableWidgetItem>
#include <QDir>
#include <QFile>
#include <QTextStream>
#include <QProcess>
#include <QStringList>
#include <QString>
#include <QMessageBox>
#include <QHeaderView>
#include <QTableWidget>
#include <QTableWidgetItem>
#include <QFileDialog>
#include <QListWidgetItem>
#include <QLabel>
#include <QPixmap>
#include <QDebug> 
  LaAnimalManager::LaAnimalManager(QWidget* parent, Qt::WFlags fl)
: QDialog(parent,fl)
{
    //required by Qt4 to initialise the ui
  setupUi(this);
  readSettings();
  lblAnimalPix->setScaledContents(true);
  connect(tblAnimals, SIGNAL(cellClicked( int,int)),
      this, SLOT(cellClicked( int,int)));
    //connect(pbnAnimalPic, SIGNAL(clicked()), this, SLOT(on_pbnAnimalPic_clicked()));

  refreshAnimalTable();
    //disable these buttons unless experimental is allowed
  pbnImport->setVisible(false);
  pbnExport->setVisible(false);
}
//
LaAnimalManager::~LaAnimalManager()
{
  writeSettings();
}

void LaAnimalManager::readSettings()
{
  QSettings mySettings;
  QPoint pos = mySettings.value("mainwindow/pos", QPoint(200, 200)).toPoint();
  QSize size = mySettings.value("mainwindow/size", QSize(400, 400)).toSize();
  resize(size);
  move(pos);
}

void LaAnimalManager::writeSettings()
{
  QSettings mySettings;
  mySettings.setValue("mainwindow/pos", pos());
  mySettings.setValue("mainwindow/size", size());
}

void LaAnimalManager::refreshAnimalTable(QString theGuid)
{

  mAnimalMap.clear();
  tblAnimals->clear();
  tblAnimals->setRowCount(0);
  tblAnimals->setColumnCount(2);


    //we do this in two passes
    //in the first pass we populate a qmap with all the layersets
    //we find....
  mAnimalMap = LaUtils::getAvailableAnimals();

    //the second pass populates the table
    //doing it from the map ensures that the rows
    //are sorted by layerset name

  int mySelectedRow=0;
  int myCurrentRow=0;
  QMapIterator<QString, LaAnimal> myIterator(mAnimalMap);
  while (myIterator.hasNext())
  {
    myIterator.next();
    LaAnimal myAnimal = myIterator.value();
   // qDebug()JASONDIDTHIS<< myAnimal.toText();
    if (theGuid.isEmpty())
    {
        //qDebug("No default active row was requested.Assigning to myAnimal.guid()!");;
      theGuid=myAnimal.guid();
    }
    if (myAnimal.guid()==theGuid)
    {
      mySelectedRow=myCurrentRow;
    }
      // Insert new row ready to fill with details
    tblAnimals->insertRow(myCurrentRow);
    QString myGuid = myAnimal.guid();
    qDebug() << "Inserting animal with guid: " << myGuid;
      // Add details to the new row
    QTableWidgetItem *mypFileNameItem= new QTableWidgetItem(myGuid);
    tblAnimals->setItem(myCurrentRow, 0, mypFileNameItem);
    QTableWidgetItem *mypNameItem = new QTableWidgetItem(myAnimal.name());
    tblAnimals->setItem(myCurrentRow, 1, mypNameItem);
      //display an icon indicating if the layerset is local or remote (e.g. terralib)
      //LaAnimal::Origin myOrigin = myAnimal.origin();
      //if (myOrigin==LaAnimal::USERPROFILE)
      //{
    QIcon myIcon;
    myIcon.addFile(":/localdata.png");
    mypNameItem->setIcon(myIcon);
      //}
      //else if (myOrigin==LaAnimal::ADAPTERPROFILE)
      //{
      //QIcon myIcon;
      //myIcon.addFile(":/remotedata.png");
      //mypNameItem->setIcon(myIcon);
      //}
      //else if (myOrigin==LaAnimal::UNDEFINED)
      //{
      //  mypNameItem->setTextColor(Qt::yellow);
      //}
    myCurrentRow++;
  }

  if (myCurrentRow>0)
  {
    tblAnimals->setCurrentCell(mySelectedRow,1);
    cellClicked(mySelectedRow,1);
  }
  else
  {
    on_toolNew_clicked();
  }
  QStringList headerLabels;
  headerLabels << "File Name" << "Name";
  tblAnimals->setHorizontalHeaderLabels(headerLabels);
  tblAnimals->setColumnWidth(0,0);
  tblAnimals->setColumnWidth(1,tblAnimals->width());
  tblAnimals->horizontalHeader()->hide();
  tblAnimals->verticalHeader()->hide();
  tblAnimals->horizontalHeader()->setResizeMode(1,QHeaderView::Stretch);
}

void LaAnimalManager::cellClicked(int theRow, int theColumn)
{
    //note we use the alg name not the id becuase user may have customised params
 // qDebug()JASONDIDTHIS<< "LaAnimalManager::cellClicked";
  QString myGuid = tblAnimals->item(tblAnimals->currentRow(),0)->text();
 // qDebug()JASONDIDTHIS<< "Guid is: " << myGuid;
  QString myFileName = myGuid + ".xml";
  selectAnimal(myFileName);
}
void LaAnimalManager::selectAnimal(QString theFileName)
{
 // qDebug()JASONDIDTHIS<< "selectAnimal Called : " << theFileName;
  QString myAnimalDir = LaUtils::userAnimalProfilesDirPath();
  LaAnimal myAnimal;
  myAnimal.fromXmlFile(myAnimalDir + QDir::separator() + theFileName);
  mAnimal=myAnimal;
  showAnimal();
}

void LaAnimalManager::showAnimal()
{
  leName->setText(mAnimal.name());
  leDescription->setText(mAnimal.description());
  sbMeatFoodValue->setValue(mAnimal.meatFoodValue());
  sbUsableMeatPercent->setValue(mAnimal.usableMeat());
  sbKillWeight->setValue(mAnimal.killWeight());
  sbAdultWeight->setValue(mAnimal.adultWeight());
  sbConceptionEfficiency->setValue(mAnimal.conceptionEfficiency());
  sbFemalesToMales->setValue(mAnimal.femalesPerMale());
  sbGrowTime->setValue(mAnimal.growTime());
  sbDeathRate->setValue(mAnimal.deathRate());
  cbFeedEnergyType->setCurrentIndex(mAnimal.feedEnergyType());
  sbEnergyForPregnant->setValue(mAnimal.gestating());
  sbEnergyForLactating->setValue(mAnimal.lactating());
  sbEnergyForMaintenance->setValue(mAnimal.maintenance());
  sbEnergyForJuvenilePerKg->setValue(mAnimal.juvenile());
  sbSexualMaturity->setValue(mAnimal.sexualMaturity());
  sbBreedingLife->setValue(mAnimal.breedingExpectancy());
  sbYoungPerBirth->setValue(mAnimal.youngPerBirth());
  sbWeaningAge->setValue(mAnimal.weaningAge());
  sbWeaningWeight->setValue(mAnimal.weaningWeight());
  sbGestationTime->setValue(mAnimal.gestationTime());
  sbEstrousCycleTime->setValue(mAnimal.estrousCycle());
  sbLactationTime->setValue(mAnimal.lactationTime());
  checkBoxMilk->setChecked(mAnimal.milk());
  sbMilk->setValue(mAnimal.milkGramsPerDay());
  sbMilkFoodValue->setValue(mAnimal.milkFoodValue());
  checkBoxFleece->setChecked(mAnimal.fleece());
  sbFleeceWeight->setValue(mAnimal.fleeceWeightKg());
  lblAnimalPix->setPixmap(mAnimal.imageFile());
}

void LaAnimalManager::on_pbnAnimalPic_clicked()
{
  LaUtils myUtils;
  QString myFile = myUtils.openGraphicFile();
  lblAnimalPix->setPixmap(myFile);
  mImageFile = myFile;
}

void LaAnimalManager::on_pushButtonLoad_clicked()
{
    //
  mAnimal.fromXmlFile("/tmp/animal.xml");
  showAnimal();
}

void LaAnimalManager::on_pushButtonSave_clicked()
{
    // not yet implemented
}

void LaAnimalManager::on_toolNew_clicked()
{
  qDebug("New toolbutton clicked");
  LaAnimal myAnimal;
  myAnimal.setGuid();
  mAnimal = myAnimal;
  showAnimal();
}

void LaAnimalManager::resizeEvent ( QResizeEvent * theEvent )
{
  tblAnimals->setColumnWidth(0,0);
  tblAnimals->setColumnWidth(1,tblAnimals->width());
  tblAnimals->horizontalHeader()->setResizeMode(1,QHeaderView::Stretch);
}

void LaAnimalManager::on_toolCopy_clicked()
{
  qDebug("Copy toolbutton clicked");
  if (tblAnimals->currentRow() < 0)
  {
    return;
  }
    //to clone, we get the algorithm guid that is currently selected
  QString myGuid = tblAnimals->item(tblAnimals->currentRow(),0)->text();
  if (myGuid.isEmpty())
  {
    return;
  }
  QString myOriginalFileName = LaUtils::userAnimalProfilesDirPath() + QDir::separator() + myGuid + ".xml";
  LaAnimal myAnimal;
  myAnimal.fromXmlFile(myOriginalFileName);
  /*
  int myCount = 1;
  while (mAnimal.contains(myProfileName))
  {
    myProfileName = tr("Copy ") + QString::number(myCount++) + " of " + myAnimal.name();
  }
  */
    //assign this layerset its own guid
  myAnimal.setGuid();
  QString myNewFileName = LaUtils::userAnimalProfilesDirPath() + QDir::separator() + myAnimal.guid() + ".xml";
  myAnimal.setName(tr("Copy of ") + myAnimal.name());
  myAnimal.toXmlFile(myNewFileName);
  refreshAnimalTable(myAnimal.guid());
}
void LaAnimalManager::on_toolDelete_clicked()
{
  qDebug("Delete toolbutton clicked");
  if (tblAnimals->currentRow() < 0)
  {
    return;
  }
  QString myGuid = tblAnimals->item(tblAnimals->currentRow(),0)->text();
  if (!myGuid.isEmpty())
  {
    QFile myFile(LaUtils::userAnimalProfilesDirPath() + QDir::separator() + myGuid + ".xml");
    if (!myFile.remove())
    {
      QMessageBox::warning(this, tr("Landuse Analyst"),
      tr("Unable to delete file \n") + myFile.fileName());
    }
    refreshAnimalTable();
  }
}
void LaAnimalManager::on_pbnApply_clicked()
{
  mAnimal.setName(leName->text());
  mAnimal.setDescription(leDescription->text());
  mAnimal.setMeatFoodValue(sbMeatFoodValue->value());
  mAnimal.setUsableMeat(sbUsableMeatPercent->value());
  mAnimal.setKillWeight(sbKillWeight->value());
  mAnimal.setAdultWeight(sbAdultWeight->value());
  mAnimal.setConceptionEfficiency(sbConceptionEfficiency->value());
  mAnimal.setFemalesToMales(sbFemalesToMales->value());
  mAnimal.setGrowTime(sbGrowTime->value());
  mAnimal.setDeathRate(sbDeathRate->value());
  
  QString mySelectedFeedEnergyType = QString(cbFeedEnergyType->currentText());
  EnergyType myFeedEnergyType;

  if (mySelectedFeedEnergyType == "KCalories")
  {
    myFeedEnergyType = KCalories;
    mAnimal.setFeedEnergyType(myFeedEnergyType);
  }
  else if (mySelectedFeedEnergyType == "TDN")
  {
    myFeedEnergyType = TDN;
    mAnimal.setFeedEnergyType(myFeedEnergyType);
  }
  
  mAnimal.setGestating(sbEnergyForPregnant->value());
  mAnimal.setLactating(sbEnergyForLactating->value());
  mAnimal.setMaintenance(sbEnergyForMaintenance->value());
  mAnimal.setJuvenile(sbEnergyForJuvenilePerKg->value());
  mAnimal.setSexualMaturity(sbSexualMaturity->value());
  mAnimal.setBreedingExpectancy(sbBreedingLife->value());
  mAnimal.setYoungPerBirth(sbYoungPerBirth->value());
  mAnimal.setWeaningAge(sbWeaningAge->value());
  mAnimal.setWeaningWeight(sbWeaningWeight->value());
  mAnimal.setGestationTime(sbGestationTime->value());
  mAnimal.setEstrousCycle(sbEstrousCycleTime->value());
  mAnimal.setLactationTime(sbLactationTime->value());
  mAnimal.setMilk(checkBoxMilk->checkState());
  mAnimal.setMilkGramsPerDay(sbMilk->value());
  mAnimal.setMilkFoodValue(sbMilkFoodValue->value());
  mAnimal.setFleece(checkBoxFleece->checkState());
  mAnimal.setFleeceWeightKg(sbFleeceWeight->value());
  mAnimal.setImageFile(mImageFile);
  mAnimal.toXmlFile( LaUtils::userAnimalProfilesDirPath() +
      QDir::separator() + mAnimal.guid() + ".xml");
  refreshAnimalTable(mAnimal.guid());
}

File: la/gui/laanimalmanager.py
from qgis.PyQt.QtCore import Qt, QModelIndex, QAbstractTableModel, QSettings
from qgis.PyQt.QtGui import QBrush, QColor, QIcon, QPixmap
from qgis.PyQt.QtWidgets import QDialog, QTreeWidgetItem, QMessageBox, QTableWidgetItem, QFileDialog, QHeaderView, QComboBox
from qgis.PyQt import uic
import os

from la.lib.laanimal import LaAnimal
from la.ui.laanimalmanagerbase import LaAnimalManagerBase
from la.lib.lautils import LaUtils
from la.lib.la import EnergyType as LaEnergyType
from la.lib.la import AreaUnits as LaAreaUnits
from la.lib.la import Priority

FORM_CLASS, _ = uic.loadUiType(
    os.path.join(
        os.path.dirname(__file__),
        '../ui/laanimalmanagerbase.ui'
    )
)

class LaAnimalTableModel(QAbstractTableModel):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.mAnimals = []

    def rowCount(self, parent=QModelIndex()):
        return len(self.mAnimals)

    def columnCount(self, parent=QModelIndex()):
        return 2

    def data(self, index, role=Qt.DisplayRole):
        if not index.isValid():
            return None

        row = index.row()
        col = index.column()
        animal = self.mAnimals[row]

        if role == Qt.DisplayRole:
            if col == 0:
                return animal.name
            elif col == 1:
                return animal.description
        elif role == Qt.BackgroundRole:
            if row % 2 == 0:
                return QBrush(QColor(240, 240, 240))
        elif role == Qt.TextAlignmentRole:
            return Qt.AlignCenter

        return None

    def headerData(self, section, orientation, role=Qt.DisplayRole):
        if role == Qt.DisplayRole:
            if orientation == Qt.Horizontal:
                if section == 0:
                    return "Name"
                elif section == 1:
                    return "Description"
            elif orientation == Qt.Vertical:
                return str(section + 1)

        return None

    def addAnimal(self, animal):
        self.beginInsertRows(QModelIndex(), len(self.mAnimals), len(self.mAnimals))
        self.mAnimals.append(animal)
        self.endInsertRows()

    def removeAnimal(self, index):
        self.beginRemoveRows(QModelIndex(), index, index)
        del self.mAnimals[index]
        self.endRemoveRows()

    def getAnimal(self, index):
        return self.mAnimals[index]


# class LaAnimalManagerBase(QDialog, FORM_CLASS):
#     """Base class for the Animal Manager dialog.

#     This class handles loading the UI. All implementation logic
#     should be in the LaAnimalManager class.
#     """
#     def __init__(self, parent=None):
#         super(LaAnimalManagerBase, self).__init__(parent)
#         self.setupUi(self)

#         # Basic UI initialization
#         self.lblAnimalPix.setScaledContents(True)

#         # Set up table headers
#         self.tblAnimals.horizontalHeader().hide()
#         self.tblAnimals.verticalHeader().hide()


class LaAnimalManager(LaAnimalManagerBase):
    """The Animal Manager dialog.

    This class provides the implementation for managing animals, including
    creating, editing, copying, and deleting animals.
    """
    def __init__(self, theAnimalsMap=None, parent=None):
        # Initialize UI from base class first
        super(LaAnimalManager, self).__init__(parent)

        # Store a working copy of the animals map
        self.mAnimalsMap = theAnimalsMap.copy() if theAnimalsMap else {}

        # Initialize other variables
        self.mImageFile = ""
        self.mAnimal = LaAnimal()

        # Connect signals/slots
        self.connectSignalsSlots()

        # Populate the table with animals
        self.refreshAnimalTable()

        # Read window settings
        self.readSettings()

    def connectSignalsSlots(self):
        """Connect signals to slots for UI interaction."""
        self.tblAnimals.cellClicked.connect(self.cellClicked)
        self.toolNew.clicked.connect(self.on_toolNew_clicked)
        self.toolCopy.clicked.connect(self.on_toolCopy_clicked)
        self.toolDelete.clicked.connect(self.on_toolDelete_clicked)
        self.pbnApply.clicked.connect(self.on_pbnApply_clicked)
        self.pbnAnimalPic.clicked.connect(self.on_pbnAnimalPic_clicked)
        self.tblAnimals.itemSelectionChanged.connect(self.on_tblAnimals_itemSelectionChanged)

    def readSettings(self):
        """Read window settings from QSettings."""
        settings = QSettings()
        pos = settings.value("AnimalManager/pos", None)
        size = settings.value("AnimalManager/size", None)

        if pos is not None:
            self.move(pos)
        if size is not None:
            self.resize(size)

    def writeSettings(self):
        """Save window settings to QSettings."""
        settings = QSettings()
        settings.setValue("AnimalManager/pos", self.pos())
        settings.setValue("AnimalManager/size", self.size())

    def refreshAnimalTable(self, theGuid=None):
        """Refresh the table of animals.

        Args:
            theGuid (str, optional): GUID of the animal to select.
        """
        # Clear the table
        self.tblAnimals.clear()
        self.tblAnimals.setRowCount(0)
        self.tblAnimals.setColumnCount(2)

        # Set the headers
        headerLabels = ["File Name", "Name"]
        self.tblAnimals.setHorizontalHeaderLabels(headerLabels)
        self.tblAnimals.setColumnWidth(0, 0)  # Hide first column
        self.tblAnimals.setColumnWidth(1, self.tblAnimals.width())
        self.tblAnimals.horizontalHeader().hide()
        self.tblAnimals.verticalHeader().hide()
        self.tblAnimals.horizontalHeader().setSectionResizeMode(1, QHeaderView.Stretch)

        # Get animals
        animals_map = LaUtils.getAvailableAnimals()

        # Populate the table
        mySelectedRow = 0
        myCurrentRow = 0
        for guid, animal in animals_map.items():
            if theGuid is None and myCurrentRow == 0:  # Default to first item
                theGuid = guid

            # Select the row if the GUID matches
            if animal.guid == theGuid:  # Call guid() method
                mySelectedRow = myCurrentRow

            # Add row
            self.tblAnimals.insertRow(myCurrentRow)

            # Add GUID (hidden)
            mypFileNameItem = QTableWidgetItem(str(animal.guid))  # Call guid() method
            self.tblAnimals.setItem(myCurrentRow, 0, mypFileNameItem)

            # Add name and description
            mypNameItem = QTableWidgetItem(f"{animal.name} ({animal.description})")
            self.tblAnimals.setItem(myCurrentRow, 1, mypNameItem)

            # Add icon
            myIcon = QIcon()
            myIcon.addFile(":/localdata.png")
            mypNameItem.setIcon(myIcon)

            myCurrentRow += 1

        # Select a row if there are any
        if myCurrentRow > 0:
            self.tblAnimals.setCurrentCell(mySelectedRow, 1)
            self.cellClicked(mySelectedRow, 1)
        else:
            self.on_toolNew_clicked()

    def cellClicked(self, row, column):
        """Handle cell clicked event."""
        myGuid = self.tblAnimals.item(row, 0).text()
        self.selectAnimal(f"{myGuid}.xml")

        # Update the image if available
        if hasattr(self, 'mAnimal') and self.mAnimal and hasattr(self.mAnimal, 'imageFile') and self.mAnimal.imageFile:
            image_path = LaUtils.resolvePath(str(self.mAnimal.imageFile), 'image')
            if os.path.exists(image_path):
                pixmap = QPixmap(image_path)
                if not pixmap.isNull():
                    self.lblAnimalPix.setPixmap(pixmap)
                else:
                    self.lblAnimalPix.clear()
            else:
                self.lblAnimalPix.clear()

    def selectAnimal(self, theFileName):
        """Load an animal from a file."""
        myAnimalDir = LaUtils.userAnimalProfilesDirPath()
        filePath = os.path.join(myAnimalDir, theFileName)
        myAnimal = LaAnimal()

        if os.path.exists(filePath):
            myAnimal.fromXmlFile(filePath)
            self.mAnimal = myAnimal
            self.showAnimal()
        else:
            LaUtils.debug.log(f"Error: File does not exist: {filePath}")

    def showAnimal(self):
        """Display animal in the form."""
        if not self.mAnimal:
            return

        # Basic info
        self.leName.setText(self.mAnimal.name)
        self.leDescription.setText(self.mAnimal.description)

        # Meat production
        self.sbMeatFoodValue.setValue(self.mAnimal.meatFoodValue)
        self.sbUsableMeatPercent.setValue(self.mAnimal.usableMeat)
        self.sbKillWeight.setValue(self.mAnimal.killWeight)
        self.sbAdultWeight.setValue(self.mAnimal.adultWeight)
        self.sbConceptionEfficiency.setValue(self.mAnimal.conceptionEfficiency)
        self.sbFemalesToMales.setValue(self.mAnimal.femalesPerMale)
        self.sbGrowTime.setValue(self.mAnimal.growTime)
        self.sbDeathRate.setValue(self.mAnimal.deathRate)

        # Handle feed energy type - get the actual enum value
        energy_type = self.mAnimal.mFeedEnergyType  # Access the internal value directly
        try:
            if isinstance(energy_type, LaEnergyType):
                energy_index = energy_type.value
            else:
                energy_index = 0  # Default to first item
        except (ValueError, TypeError):
            energy_index = 0

        energy_index = max(0, min(self.cbFeedEnergyType.count() - 1, energy_index))
        self.cbFeedEnergyType.setCurrentIndex(energy_index)

        # Energy values
        self.sbEnergyForPregnant.setValue(self.mAnimal.gestating)
        self.sbEnergyForLactating.setValue(self.mAnimal.lactating)
        self.sbEnergyForMaintenance.setValue(self.mAnimal.maintenance)
        self.sbEnergyForJuvenilePerKg.setValue(self.mAnimal.juvenile)

        # Reproduction
        self.sbSexualMaturity.setValue(self.mAnimal.sexualMaturity)
        self.sbBreedingLife.setValue(self.mAnimal.breedingExpectancy)
        self.sbYoungPerBirth.setValue(self.mAnimal.youngPerBirth)
        self.sbWeaningAge.setValue(self.mAnimal.weaningAge)
        self.sbWeaningWeight.setValue(self.mAnimal.weaningWeight)
        self.sbGestationTime.setValue(self.mAnimal.gestationTime)
        self.sbEstrousCycleTime.setValue(self.mAnimal.estrousCycle)
        self.sbLactationTime.setValue(self.mAnimal.lactationTime)

        # By-products - Missing in original implementation
        self.checkBoxMilk.setChecked(self.mAnimal.milk)
        self.sbMilk.setValue(self.mAnimal.milkGramsPerDay)
        self.sbMilkFoodValue.setValue(self.mAnimal.milkFoodValue)
        self.checkBoxFleece.setChecked(self.mAnimal.fleece)
        self.sbFleeceWeight.setValue(int(self.mAnimal.fleeceWeightKg))

        # Update the image display
        image_file = getattr(self.mAnimal, '_imageFile', '')  # Access internal value directly
        if image_file:
            # Use LaUtils.resolveImagePath to find the image
            image_path = LaUtils.resolveImagePath(image_file)
            LaUtils.debug.log(f"Loading animal image from: {image_path}")

            if os.path.exists(image_path):
                pixmap = QPixmap(image_path)
                if not pixmap.isNull():
                    # Scale the pixmap to fit the label while maintaining aspect ratio
                    scaled_pixmap = pixmap.scaled(
                        self.lblAnimalPix.width(),
                        self.lblAnimalPix.height(),
                        Qt.KeepAspectRatio,
                        Qt.SmoothTransformation
                    )
                    self.lblAnimalPix.setPixmap(scaled_pixmap)
                    LaUtils.debug.log(f"Successfully loaded and displayed image: {image_path}")
                else:
                    LaUtils.debug.log(f"Failed to load image into pixmap: {image_path}")
                    self.lblAnimalPix.clear()
            else:
                LaUtils.debug.log(f"Image file not found: {image_path}")
                self.lblAnimalPix.clear()
        else:
            LaUtils.debug.log("No image file specified for animal")
            self.lblAnimalPix.clear()

    def on_toolNew_clicked(self):
        """Create a new animal."""
        myAnimal = LaAnimal()
        myAnimal.setGuid(None)  # Generate new GUID
        self.mAnimal = myAnimal
        self.showAnimal()

    def on_toolCopy_clicked(self):
        """Copy the selected animal."""
        if not self.mAnimal:
            return

        myAnimal = LaAnimal(self.mAnimal)
        myAnimal.setGuid(None)  # Generate new GUID
        self.mAnimal = myAnimal
        self.showAnimal()

    def on_toolDelete_clicked(self):
        """Delete the selected animal."""
        if not self.mAnimal:
            return

        # Confirm deletion
        reply = QMessageBox.question(
            self,
            "Confirm Delete",
            f"Are you sure you want to delete {self.mAnimal.name}?",
            QMessageBox.Yes | QMessageBox.No,
            QMessageBox.No
        )

        if reply == QMessageBox.Yes:
            # Delete the file
            myAnimalDir = LaUtils.userAnimalProfilesDirPath()
            filePath = os.path.join(myAnimalDir, f"{self.mAnimal.guid}.xml")

            try:
                os.remove(filePath)
                LaUtils.debug.log(f"Deleted animal file: {filePath}")

                # Refresh the table
                self.refreshAnimalTable()
            except Exception as e:
                LaUtils.debug.log(f"Failed to delete animal file: {str(e)}")
                QMessageBox.warning(self, "Delete Failed", f"Failed to delete animal: {str(e)}")

    def on_pbnApply_clicked(self):
        """Apply changes to the animal."""
        if not self.mAnimal:
            return

        # Update animal with form values
        self.mAnimal.name = self.leName.text()
        self.mAnimal.description = self.leDescription.text()

        # Set basic properties
        self.mAnimal.meatFoodValue = self.sbMeatFoodValue.value()
        self.mAnimal.usableMeat = self.sbUsableMeatPercent.value()
        self.mAnimal.killWeight = self.sbKillWeight.value()
        self.mAnimal.adultWeight = self.sbAdultWeight.value()
        self.mAnimal.growTime = self.sbGrowTime.value()
        self.mAnimal.deathRate = self.sbDeathRate.value()

        # Set feed energy type using internal attribute
        self.mAnimal.mFeedEnergyType = LaEnergyType(self.cbFeedEnergyType.currentIndex())

        # Set feed energy requirements - missing in original implementation
        self.mAnimal.gestating = self.sbEnergyForPregnant.value()
        self.mAnimal.lactating = self.sbEnergyForLactating.value()
        self.mAnimal.maintenance = self.sbEnergyForMaintenance.value()
        self.mAnimal.juvenile = self.sbEnergyForJuvenilePerKg.value()

        # Set reproduction parameters
        self.mAnimal.sexualMaturity = self.sbSexualMaturity.value()
        self.mAnimal.breedingExpectancy = self.sbBreedingLife.value()
        self.mAnimal.youngPerBirth = self.sbYoungPerBirth.value()
        self.mAnimal.weaningAge = self.sbWeaningAge.value()
        self.mAnimal.weaningWeight = self.sbWeaningWeight.value()
        self.mAnimal.gestationTime = self.sbGestationTime.value()
        self.mAnimal.estrousCycle = self.sbEstrousCycleTime.value()
        self.mAnimal.lactationTime = self.sbLactationTime.value()

        # Set by-products - missing in original implementation
        self.mAnimal.milk = self.checkBoxMilk.isChecked()
        self.mAnimal.milkGramsPerDay = self.sbMilk.value()
        self.mAnimal.milkFoodValue = self.sbMilkFoodValue.value()
        self.mAnimal.fleece = self.checkBoxFleece.isChecked()
        self.mAnimal.fleeceWeightKg = self.sbFleeceWeight.value()

        # Handle image file
        if self.mImageFile:
            # Use LaUtils to handle image copying and saving
            target_image_path = LaUtils.openGraphicFile()
            if target_image_path:
                self.mAnimal.imageFile = os.path.basename(target_image_path)  # Store just the filename

        # Save animal to file
        target_file = os.path.join(LaUtils.userAnimalProfilesDirPath(), f"{self.mAnimal.guid}.xml")
        LaUtils.debug.log(f"Saving animal to: {target_file}")
        success = self.mAnimal.toXmlFile(target_file)

        if success:
            self.refreshAnimalTable(self.mAnimal.guid)
        else:
            QMessageBox.warning(self, "Landuse Analyst", f"Failed to save animal to {target_file}")
            return

    def on_pbnAnimalPic_clicked(self):
        """Select an image for the animal."""
        # Use the static method instead of creating an instance
        # This approach avoids potential issues with multiple dialogs
        imagePath = QFileDialog.getOpenFileName(
            self,
            "Select Animal Image",
            "",
            "Images (*.png *.jpg *.jpeg *.bmp *.gif)"
        )[0]  # getOpenFileName returns (filename, filter)
        
        if imagePath:
            self.mImageFile = imagePath
            
            # Display the image
            pixmap = QPixmap(imagePath)
            if not pixmap.isNull():
                scaled_pixmap = pixmap.scaled(
                    self.lblAnimalPix.width(),
                    self.lblAnimalPix.height(),
                    Qt.KeepAspectRatio,
                    Qt.SmoothTransformation
                )
                self.lblAnimalPix.setPixmap(scaled_pixmap)
                LaUtils.debug.log(f"Selected image: {imagePath}")

    def on_tblAnimals_itemSelectionChanged(self):
        """Handle selection change in the animals table."""
        # Get selected row
        selectedRows = self.tblAnimals.selectedItems()
        if selectedRows:
            selectedRow = selectedRows[0].row()
            self.cellClicked(selectedRow, 1)

    def resizeEvent(self, event):
        """Handle resize event to adjust table columns."""
        super(LaAnimalManager, self).resizeEvent(event)
        # Adjust column widths
        self.tblAnimals.setColumnWidth(0, 0)  # Hide first column
        self.tblAnimals.setColumnWidth(1, self.tblAnimals.width())